conn_http.go

```diff
@@ -30,102 +35,178 @@ import (
 )
 
 func dialHttp(ctx context.Context, addr string, num int, opt *Options) (*httpConnect, error) {
-	url := &url.URL{
-		Scheme: "http",
+	u := &url.URL{
+		Scheme: opt.Scheme,
 		Host:   addr,
 	}
 
-	if opt.TLS != nil {
-		url.Scheme = "https"
+	query := u.Query()
+	query.Set("default_format", "Native")
+	u.RawQuery = query.Encode()
+
+	t := &http.Transport{
+		DialContext: (&net.Dialer{
+			Timeout:   opt.DialTimeout,
+			KeepAlive: opt.ConnMaxLifetime,
+		}).DialContext,
+		MaxIdleConns:          1,
+		IdleConnTimeout:       opt.ConnMaxLifetime,
+		ResponseHeaderTimeout: opt.ReadTimeout,
+		TLSClientConfig:       opt.TLS,
 	}
 
-	connect := &httpConnect{
+	conn := &httpConnect{
 		opt: opt,
-		transport: &http.Transport{
-			DialContext: (&net.Dialer{
-				Timeout:   opt.DialTimeout,
-				KeepAlive: opt.ConnMaxLifetime,
-			}).DialContext,
-			MaxIdleConns:          1,
-			IdleConnTimeout:       opt.ConnMaxLifetime,
-			ResponseHeaderTimeout: opt.ReadTimeout,
-			TLSClientConfig:       opt.TLS,
+		client: &http.Client{
+			Transport: t,
 		},
-		url:      url,
-		location: time.UTC, // TODO: make configurable
+		url: u,
 	}
 
-	if err := connect.ping(ctx); err != nil {
+	rows, err := conn.query(ctx, func(*connect, error) {}, "SELECT timeZone()")
+	if err != nil {
 		return nil, err
 	}
 
-	return connect, nil
+	for rows.Next() {
+		var serverLocation string
+		rows.Scan(&serverLocation)
+		location, err := time.LoadLocation(serverLocation)
+		if err != nil {
+			return nil, err
+		}
+		conn.location = location
+	}
+
+	return conn, nil
 }
 
 type httpConnect struct {
-	opt       *Options
-	url       *url.URL
-	transport *http.Transport
-	location  *time.Location
+	opt      *Options
+	url      *url.URL
+	client   *http.Client
+	location *time.Location
+}
+
+func (h *httpConnect) isBad() bool {
+	if h.client == nil {
+		return true
+	}
+	return false
+}
+
+func writeData(encoder *binary.Encoder, block *proto.Block) error {
```

Comment:
these two result in a new encoder and decoder being created each time? any reason we can't create and hold on the connection like std interface? seems like an unnecessary overhead

Reply:
We cannot change reader and writer in Decoder and Encoder. 
So in batch, we need encoder only once to create a body for a request.
In query, we read a body from a response, so sources are changed each request.

Maybe I don't see some option

Reply:
Can we not create the Encoder and Decoder once, store in the connection and reuse across requests?

Reply:
I think yes, if encoder and decoder will be extended with `Reset(input)` method. So we can reset `Writer` and `Reader` inside encoder and decoder, respectively

Reply:
See https://github.com/ClickHouse/clickhouse-go/pull/622/commits/dbd3b95c082962057a7557c2cc52b099ce99d625